proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,223 @@
+* Proposal: [0010](0010-resource-element-type-validation.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: Joshua Batista
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+* Issues: [#75676](https://github.com/llvm/llvm-project/issues/75676)
+
+## Introduction
+Resources are often used in HLSL, with various resource element types (RETs).
+
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+In this code, the RET is `float`, and the resource type is `RWBuffer`.
+All resources can be placed in two categories, raw buffers and typed buffers. 
+Below is a description of all resources and their corresponding categories
+* raw buffers
+  * [Append|Consume|RW|RasterizerOrdered]StructuredBuffer
+  * [RW]ByteAddressBuffer
+* typed buffers
+  * [RW|RasterizerOrdered]Buffer
+  * [Feedback]Texture*
+
+There is a distinct set of rules that define valid RETs for raw buffer resources, 
+and a separate set of rules that define valid RETs for typed buffer resources.
+
+RETs for typed buffer resources:
+* Are not intangible (e.g., isn't a resource type)
+* Must be vectors or scalars of arithmetic types, not bools nor enums nor arrays
+* The type should be line-vector layout compatible (homogenous, at most 4 elements, and at most 128 bits in size) 

```

Comment:
I have always had a problem with this terminology.
- It's inventing new terminology not connected to any prior terminology used for the concept to which it applies: "line-vector"?  It's trying to sound more general than it actually is.
- If you were to think about what it could/should be referring to, you could justifiably think it's about what fits into one constant buffer row (line?) in the legacy (current) constant buffer layout.  But that would be a mistake, since the rules are different.  (for example, you can put 8 native 16-bit values or a struct containing non-homogenous types in one row)
- It implies that this type is used in the (memory) layout of the resource itself.  It's not: the element stored and the layout is dependent on runtime factors: the DXGI resource type and device-dependent factors.  The element type in HLSL for a typed resource is just how you want to interpret the values loaded from that up-to-4-component typed resource element.
- The description sounds like it would support non-homogenous types (a struct with several different types that fit should work, as it would for constant buffers)
- This 128-bit restriction sounds equivalent, but hides the true meaning: each 64-bit component in HLSL is translated to 2-32-bit components for the resource element type, with a bitcast for translation.  This has implications for the actual element type used, compatible DXGI type mapping, and allowed operations.  This has a limit of 2 64-bit components because those 2 components expand into a 4 component element type, not because it's 128 bits in size.
- We could probably create specializations that handle 64-bit type translation directly to/from the translated element type in our HLSL header, rather than adding special-case code to the compiler.  In this scenario, the real element type for the resource does not use a 64-bit type as the component type, but uses the translated element type with 32-bit uint components instead.

I would rather a more specific term be used for the more specific typed resource element constraint.

How about "typed-element compatible" and `__builtin_hlsl_typed_element_compatible`?
- A "typed-element" flows from existing terminology: typed resources are textures and buffers with a DXGI type applying to the resource view descriptor at runtime.  These support translation upon load/store to the HLSL element type used in a shader (depending on the types at either end and the operation being performed).
- The element must be a scalar or vector of an integer or floating point type with a maximum of 4 components.  All loads and stores are performed at element granularity, which is a 4 component vector (or struct) in the shader IR, whether or not all those components are read/defined.
- If a 64-bit component type is used in HLSL, each 64-bit value maps to 2-32-bit `uint` components which will be reinterpreted together as the 64-bit value in HLSL.  This has implications for supported operations, and the maximum number of 64-bit components allowed.  `double2` or `[u]int64_t2` becomes `uint4`, which is the maximum number of components, and because the real component used is integer, it triggers those operation restrictions as well.

These above rules and description would apply to this "typed-element compatible"/`__builtin_hlsl_typed_element_compatible` constraint.

The short description could be:
- slightly modified: "at most 4 homogenous scalars, and at most 128 bits in total size" (avoiding using "element" for scalar/component since the whole thing is the element type)
- or "at most 4 homogenous scalars after translation of 64-bit scalars into pairs of uint32_t scalars".  This makes the implied translation clear right in the short constraint description.
- or subsume the scalar or vector and component type rule:
  "scalar or vector of a floating-point or integer type, with a maximum of 4 components after translating 64-bit components into pairs of `uint32_t` components"

There some additional constraints on the element type relating to the way the resource is used (methods called), but these would need to be enforced separately on the method calls:
- If comparison sample (`SampleCmp*`) methods are used, the element type in HLSL must be a scalar floating point type of at most 32-bits.
- If other `Sample` methods are used, integer components are only allowed in the element type when the shader model is 6.7 or greater.  This is triggered by the use of double components because they break down into 2 `uint32_t` components.
There are additional runtime restrictions for legal type combinations between the shader and the DXGI type, but these cannot be enforced by the compiler when they depend on runtime state.  For instance: typed UAVs using a unorm or snorm DXGI type need to have a matching unorm float or snorm float component type in the HLSL resource element declaration.

Reply:
The naming used here is intended to be consistent with https://github.com/microsoft/hlsl-specs/pull/321. Please comment on that PR if you think we need to adjust the naming there.